### PR TITLE
[MAINTENANCE] fix get available data assets names for fds

### PR DESCRIPTION
**Pull request by jshaikGX**
Changes proposed in this pull request:
- get_available_data_asset_names() did not return fds asset names before. This fix here changes that

### Definition of Done
Please delete options that are not relevant.

- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -3917,27 +3917,28 @@ def get_available_data_asset_names(
             ):  # Iterate over both together
                 for idx, datasource_name in enumerate(datasource_names):
                     datasource = self.get_datasource(datasource_name)
-                    assert not isinstance(
```
Comment (alexsherstinsky) [inline great_expectations/data_context/data_context/abstract_data_context.py:3920]: @jshaikGX Before we can confirm this approach, we need to make sure that we are not breaking anything.  The problem is that the existing implementation is based on GX V2!  So the question is -- will anything break that we are still supposed to support?  Perhaps something needs to be done to ensure that whoever is relying on the V2 construct `batch_kwargs_generator_names` (or anything to do with `batch_kwargs_generator`) should not be.  Maybe some kind of a warning or a deprecation notice is in order?  We never supported this call `get_available_data_asset_names()` at the DataContext level properly in V3 (only in Datasource) which is what you are using, but this will leave the V2 implementation non-existent.  Recently, @anthonyburdi deleted quite a bit of V2 functionality, please add him as a reviewer.  Thank you.

Comment (jshaikGX) [reply great_expectations/data_context/data_context/abstract_data_context.py:3920]: @anthonyburdi

File: tests/data_context/test_data_context_datasources.py [hunk 2]
```diff
@@ -453,3 +453,34 @@ def test_list_datasources() -> None:
...
+                    "class_name": "RuntimeDataConnector",
+                    "module_name": "great_expectations.datasource.data_connector",
+                    "batch_identifiers": ["default_identifier_name"],
+                },
+            },
+        }
+    }
+    context = gx.get_context(project_config=project_config)
+
+    datasource_name = "my_experimental_datasource_awaiting_migration"
```
Comment (NathanFarmer) [inline tests/data_context/test_data_context_datasources.py]: I would just call this `my_fluent_pandas_datasource`.

File: tests/data_context/test_data_context_datasources.py [hunk 3]
```diff
@@ -453,3 +453,34 @@ def test_list_datasources() -> None:
...
+    context.sources.add_pandas(datasource_name)
```
Comment (NathanFarmer) [inline tests/data_context/test_data_context_datasources.py]: This just creates a datasource for you. You need to also add an asset like:
```
datasource = context.sources.add_pandas(datasource_name)
data_asset = datasource.add_dataframe_asset(df)
```

File: tests/data_context/test_data_context_datasources.py [hunk 4]
```diff
@@ -453,3 +453,34 @@ def test_list_datasources() -> None:
...
+
+    assert len(context.get_available_data_asset_names()) == 2
```
Comment (NathanFarmer) [inline tests/data_context/test_data_context_datasources.py]: What happens when you pass `datasource_names`? Do you get the same behavior for the block-config datasource as the fluent datasource?

**Review by NathanFarmer — CHANGES_REQUESTED** [review]

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py:3920]: I haven't yet looked at the code in this PR, but just to respond to this comment - yes we have deprecated the V2 api several minor versions ago and have started to delete some of the code per our deprecation policy (with the rest being deleted hopefully soon).

Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py:3920]: Also, I would suggest adding the `deprecated_argument` decorator to this method for `batch_kwargs_generator_names`, looks like we missed adding that when adding this method to the public API.

Comment (anthonyburdi) [reply great_expectations/data_context/data_context/abstract_data_context.py:3920]: Also, this does not look like it will break existing v2 api workflows yet, there are isinstance checks for `FluentDatasource`.

**Review by anthonyburdi — APPROVED** [review]
This looks like it will work. The get_available_data_asset_names() method is a bit hard to follow with all the nesting, I am going to approve but if you are able to refactor I think it could be made easier to separate the v2 / v3 / fluent logic so that we can remove it later.

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 5]
```diff
@@ -3917,27 +3917,28 @@ def get_available_data_asset_names(
...
-                        batch_kwargs_generator_names[idx]
-                    )
+                    if isinstance(datasource, FluentDatasource):
+                        data_asset_names[datasource_name] = datasource.get_asset_names()
+                    else:
+                        data_asset_names[
+                            datasource_name
+                        ] = datasource.get_available_data_asset_names(
+                            batch_kwargs_generator_names[idx]
+                        )
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:3933]:
```
great_expectations/data_context/data_context/abstract_data_context.py:3925: error:
Incompatible types in assignment (expression has type
"Union[Dict[str, List[str]], Any]", target has type "Set[str]")  [assignment]
                            ] = datasource.get_available_data_asset_names(
                                ^
  ```
You are assigning a `Set[str]` to a `List[str]`. Easiest fix here is to convert the `Set` to a `List`.
```python
data_asset_names[
     datasource_name
    ] = list(datasource.get_available_data_asset_names(
     batch_kwargs_generator_names[idx]
                        ))
```

**Review by NathanFarmer — APPROVED** [review]

File: tests/data_context/test_data_context_datasources.py [hunk 6]
```diff
@@ -453,3 +453,21 @@ def test_list_datasources() -> None:
     context.sources.add_pandas(datasource_name)
 
     assert len(context.list_datasources()) == 2
+
+
+@pytest.mark.unit
+def test_get_available_data_assets_names(test_df_pandas, empty_data_context) -> None:
```
Comment (cdkini) [inline tests/data_context/test_data_context_datasources.py]: As `empty_data_context` scaffolds the filesystem as part of setup, I wouldn't consider this a unit test but rather an integration test - getting true, pure unit tests with the context is difficult work

File: tests/data_context/test_data_context_datasources.py [hunk 7]
```diff
@@ -453,3 +453,21 @@ def test_list_datasources() -> None:
...
+
+    assert len(empty_data_context.get_available_data_asset_names()) == 1
+
+    data_asset_names = dict(
+        empty_data_context.get_available_data_asset_names(
+            datasource_names=datasource_name
+        )
+    )
+
+    assert "test_data_frame" in str(data_asset_names.values())
```
Comment (cdkini) [inline tests/data_context/test_data_context_datasources.py]: Nit - instead of using two separate strings, can we do something like this:

```
asset_name = "test_data_frame"
...
datasource.add_dataframe_asset(name=asset_name)
...
assert asset_name in data_asset_names.values()
```

Comment (cdkini) [reply tests/data_context/test_data_context_datasources.py]: also this casting to a string on L473 shouldn't be necessary

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 8]
```diff
@@ -3897,7 +3897,7 @@ def get_available_data_asset_names(
         Raises:
             ValueError: `datasource_names` is not None, a string, or list of strings.
         """
-        data_asset_names = {}
+        data_asset_names: dict[str, list[str]] = {}
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Could we annotate the return value on L3886? Do we know what to expect now?

Our typing of the OSS codebase is gradual so let's chip away at it if its not too big of a lift

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 9]
```diff
@@ -3917,27 +3917,31 @@ def get_available_data_asset_names(
...
-                    ), 'Method "get_available_data_asset_names" not implemented for FluentDatasource'
-                    data_asset_names[
-                        datasource_name
-                    ] = datasource.get_available_data_asset_names(
-                        batch_kwargs_generator_names[idx]
-                    )
+                    if isinstance(datasource, FluentDatasource):
+                        data_asset_names[datasource_name] = list(
+                            datasource.get_asset_names()
+                        )
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Nit - probably not super necessary but if we're converting from a set to a list, can we use `sorted()`? It would be nice to get consistent ordering here since we're not guaranteed that in a set.

Same throughout

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 10]
```diff
@@ -3948,15 +3952,17 @@ def get_available_data_asset_names(
...
+                    else:
+                        data_asset_names[datasource_name] = list(
+                            datasource.get_available_data_asset_names()
+                        )
                 except ValueError:
                     # handle the edge case of a non-existent datasource
-                    data_asset_names[datasource_name] = {}
+                    data_asset_names[datasource_name] = list()
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py]: Nit - can we use `[]`? Either works but we tend to use this syntax in our codebases

File: great_expectations/data_context/data_context/abstract_data_context.py (hunk 1 above)
Comment (cdkini) [reply great_expectations/data_context/data_context/abstract_data_context.py:3920]: yeah this shouldn't break V2 - that being said, I don't want to add scope creep to @jshaikGX 's work here.

Are we confident in this deprecation? Should we surface in arch to confirm?

File: tests/data_context/test_data_context_datasources.py [hunk 11]
```diff
@@ -453,3 +453,22 @@ def test_list_datasources() -> None:
...
+
+    assert len(empty_data_context.get_available_data_asset_names()) == 1
+
+    data_asset_names = dict(
+        empty_data_context.get_available_data_asset_names(
+            datasource_names=datasource_name
+        )
+    )
+
+    assert asset_name in data_asset_names["my_fluent_pandas_datasource"]
```
Comment (cdkini) [inline tests/data_context/test_data_context_datasources.py]: Nit - let's use `datasource_name` as the key
